### ISSUE TITLE
Subscription Hints 

### ISSUE THREAD

**Issue by suhasHere** [moq-wg/moq-transport#245]
Subscribers needs a way to hint the publishers on when do they have the want the objects to delivered to them within the requested track. This is needed in case due to the fact that different media types can be handled differently by the application ( audio can be delivered to the application immediately,  video needs some form of key-frame to have the best experience or something more specific that the application wants )

In that effect , i want to propose an **experimental** extension to the Track Request Parameters called "SubscriptionHint" with the following variants.

* latest (0):
The "latest" subscription hint identifies delivery of the objects for the
requested track to begin from the most recent group and object sequence.

* catchup (1):
The "catchup" subscription hint identifies delivery of the objects for the
requested track from the beginning of the group provided in the GroupSequence track request param. 

* waitup(2): 
The "waitup" subscription hint identifies delivery of the objects for the
requested track  from the beginning of the new group.

**Comment by fluffy** [comment]
One of the reason this is key, is often a client is doing a subscribe to a track with a different bandwidth because there is not enough bandwidth for the current track. ( the case where a client is receiving a high bandwidth track, and wants to move a low bandwidth track of the same stuff ). In this case, there is already congestion happening and the client wants to move to the new track quickly without making the congestion worse. Hints along the lines of this really help in this critical case.

**Comment by kixelated** [comment]
Oof yeah this is a difficult problem that we need to solve.

### Goal:
* A new player can start X seconds behind the live playhead, where X is the target size of the jitter buffer.
    * Real-time will set this to 0.
* An existing player  can switch between tracks at timestamp Y. (@fluffy's example)

### Problem:
* Neither the player nor the relay knows the mapping from group <-> timestamp.
* Tracks might not have aligned groups; different offsets, frequencies, boundaries, etc.

**Comment by kixelated** [comment]
My issue with this proposal is that (0) latest and (2) waitup are based on the relay's cache and are extremely racey. The assumption is that the relay always has the latest group/object in cache. The reality is that it often won't, nor does it even know what the "latest" group is supposed to be.

For example: A player is subscribed to 720p and encounters congestion after `group=7`. It wants to subscribe to 360p and it has three choices:

### (0) SUBSCRIBE 360p latest 
* **If the latest object in cache is `group=7 object=49`**, then we receive some undecodable objects until `group=8` starts. It's wasteful but it works.
* **If the latest object in cache is `group=8 object=5`**, then `group=8` is undecodable and and playback stalls until `group=9` is generated in a few seconds (oof).
* **If there's no cache**, we could forward the SUBSCRIBE upstream with the `latest` and run into either of the two above scenarios.
* **If there's a partial cache**, the "latest" object might actually be in the past. For example, another subscriber issued a SUBSCRIBE for `group=6` 100ms ago, so our latest in cache might be `group=6 object=30`. We incorrectly start delivering at `group=6 object=30` and exacerbate congestion. (oof)


### (1) SUBSCRIBE 360p catchup 8
This works the best because it doesn't depend on what is in the relay's cache. It knows to always start at `group=8`.

However this requires groups to be aligned across different tracks (ex. HLS). If 720p `group=8` does not correspond to 360p `group=8`, then this isn't an option. It's also not an option for startup since the player doesn't know the latest group number.

### (2) SUBSCRIBE 360p waitup
* **If the latest object in cache is `group=7 object=49`**, then we correctly wait until `group=8` starts.
* **If the latest object in cache is `group=8 object=5`**, then playback stalls until  `group=9` is generated in a few seconds (oof).
* **If there's no cache**, we could forward the SUBSCRIBE upstream with the `waitup` and run into either of the two above scenarios.
* **If there's a partial cache**, the "latest" object might actually be in the past. For example, another subscriber issued a SUBSCRIBE for `group=6` 100ms ago, so our latest in cache might be `group=6 object=30`. We incorrectly start delivering at `group=7` and exacerbate congestion. (oof)

**Comment by kixelated** [comment]
That was a wall of text; hopefully I explained it well enough.

Like I said initially, the fundamental issue is that neither the player nor the relay know the mapping from group <=> timestamp. What if we resolved that?

# Proposal: Relay knows timestamp <=> group mapping
* `OBJECT` contains a timestamp field
* `SUBSCRIBE` contains a relative/absolute timestamp.
* `UNSUBSCRIBE` contains an optional timestamp for graceful termination.


This suffers from some rounding/alignment issues but it's fine. Relative timestamp requests are also problematic with partial caches.

# Proposal: Player knows timestamp <=> group mapping
* `/timeline` track contains group/timestamp pairs.
* `SUBSCRIBE` contains start group and optional end group.
* `UNSUBSCRIBE` contains optional end group for graceful termination.

This is absolutely the best option for relays. It knows what groups are missing in the cache and can fill them on request. It also means the relay can have multiple SUBSCRIBE requests upstream for the same track without fear of duplication.

The problem is that the media protocol needs to produce a live group listing kinda like HLS. This incurs an extra RTT on startup, although it's super useful for seeking backwards and DVR.

**Comment by wilaw** [comment]
I think there are scalability and flexibility benefits to keeping relays agnostic to time and in fact to any internal characteristics of the objects they are serving. Keeping them decoupled allows the content to change (and the complete timing mechanism if so desired), without having to change the relay networks.

The introduction of relay’s requiring timestamp knowledge stems from the original assertion that `Neither the player nor the relay knows the mapping from group <-> timestamp.` I think this need not be true. We can teach players the relationship between group and timestamp without too much difficulty. We do this all the time with HLS and DASH. In both cases the HTTP servers delivering these formats know nothing about the relationship between segments and time, yet we are able to build players which seek to precise points in a stream.

We can do something similar with our streaming formats and the ideas already presented in this thread by Suhas and Luke. One difference is that I would make would be to allow the group offset described by Suhas to be either an absolute or relative one. As a proposal, what if the WARP catalog listed two things:
1.	The GROUP duration in milliseconds. 
2.	The name of a track which described GROUP number/timestamp pairs (exactly the track suggested by Luke above). We assume that this timeline track allows delta updates, perhaps describing the complete DVR window every minute at a group boundary, with incremental updates in between. 

Use cases:
1.	**Start as close to the live edge as possible**
a.	Since the last group represents the last independently accessible point in the stream, the player would issue a plain SUBSCRIBE 
b.	The relay would serve content starting with the last group it has cached (its default behavior).
c. If the relay does not have this content cached, and there is no pending subscription, then it would issue its own SUBSCRIBE upstream. 
2.	**Start with a 5s buffer**
a.	The player would calculate how many group’s it would need to guarantee a 5s buffer. Assuming the advertised group duration was 2000, then it would need 3 groups. Using zeroth based relative group addressing, it would issue SUBSCRIBE (-2). 
b.	The relay would serve from the 3rd last group that it had cached. 
c. If the relay does not have the content cached, it would send its own SUSBCRIBE(-2) upstream. If it has an open subscription, but has only received 2 groups, then it should wait for the 3rd before responding. 
4.	**Seek to a wall clock time of T seconds**
a.	The player would subscribe to the /timelines track, using SUBSCRIBE so that it receives the last full DVR description.
b.	Based on the data in the timeline track, the player calculates that it needs a group number of 5321. It issues a SUBSCRIBE(5321).
c.	The relay recognizes that this is an absolute reference to a GROUP (as all relative ones are negative) and returns content starting at GROUP #5321.
5. **Player encounters congestion at group 53 on 720p and wants to switch to 360p track**
a. WARP places the packaging restriction that group numbers across tracks must carry the same media time offset (it actually has this requirement already)
b. The player knows it was playing group 53 when it encountered slowness. Its next chance for a switch will be at a group boundary. Since it knows the group duration from the catalog, it can decide whether it is better off waiting for the next group, or reloading from the same group. In this case it is near the end of the group, so it decides to fetch the next group.
c. It issues a SUBSCRIBE (54) for the 360p track and a UNSUBSCRIBE(53) for the 720p. 

Cheers
Will

**Comment by kixelated** [comment]
Absolutely @wilaw 

The main difficulty with relative requests is deduplication: a relay has trouble merging subscriptions that use absolute and relative groups when the cache is empty.

ex.
SUBSCRIBE 480p -3
(50ms delay)
SUBSCRIBE 480p 351

The relay doesn't know if the existing upstream subscription (-3) will fulfill both. In fact the relay doesn't know if 351 is even remotely close; it could be a group from minutes/hours ago. We don't want to download the same objects twice, but maybe that's okay when the cache is empty?

I think we either:
1. only use only absolute groups or
2. add a way to OR an existing SUBSCRIBE

**Comment by kixelated** [comment]
Thinking about it some more.

If the player only knows the duration of each group (fixed), then it could use relative timestamps, but it's still a massive headache for the relay.

Suppose two SUBSCRIBES from separate viewers:

```
SUBSCRIBE 0 6328
SUBSCRIBE 0 -3
```

A relay with an empty cache issues the `SUBSCRIBE 6328` upstream. It starts getting some OBJECTs: `6328 6329 6330`. And then it gets a `SUBSCRIBE -3`.

It would be an error to serve those three objects from cache. Why? Who knows if they were the latest three or from an hour ago.

The relay could hypothetically augment the existing upstream subscription: `SUBSCRIBE 0 6328 OR -3`. However, it still doesn't know which groups correspond to -3. No additional objects would be sent if 6330 actually was the live playhead, and the relay wouldn't know where to start the -3 subscribe.

IMO the relay needs a way of converting relative groups to absolute groups for this to work. Maybe SUBSCRIBE_OK contains that translation? ex. `-3 is 6345`. It's still complicated though.

**Comment by wilaw** [comment]
> Who knows if they were the latest three or from an hour ago.

You're indeed right. This the fatal weakness of the relative request approach. If the relay has the latest group cached and an active subscription, it works fine, but the player can never be sure of that and the relay can never disambiguate fresh from stale content. 

> IMO the relay needs a way of converting relative groups to absolute groups for this to work.

If the relay can do that translation, then why couldn't the player do the same and just request an absolute group to begin with?

What are some solutions to this problem?

1. Players always make absolute requests. Here are two ways to do this:
a. The catalog instructs the player in how to calculate the group number for any given time, for example, via a start time and then modulus the group duration. To avoid players having to have an accurate clock, the SETUP response from the server could echo back the current time. This is essentially a DASH approach to timing. 
b. The player subscribes to the /timeline track which describes the wallclock time and media time associated with the start of each group. This is similar to an HLS approach and has the downside of requiring 1xRTT before starting to discover the latest group. It has an advantage that it can describe groups of varying duration and 

2. Each group carries a timestamp in its header indicating the UTC time at which it was produced. This is not violating any decoupling of relays and payload because the time of production is core characteristic of any payload and is independent of the internal media time. With these timestamps in place, the relay can judge freshness. A player can issue SUBSCRIBE (-5) which means gives me this track starting with any group whose timestamp T > (now -5). We could also consider SUBSCRIBE (from: 1695571105212), which would require the player to have an accurate clock (although SETUP could be used to address that). Note that this scheme also works for DVR seeking if the player knows how to translate group number to UTC time. 

4. Same as [2] except that we add both a media timestamp and a UTC timestamp to each group. This gives the flexibility to ask relays to begin subscriptions at both media time offset and absolute and relative UTC offsets. The downsides are that it binds moq-transport to internal properties of payloads. One way round this is in core moq-transport we can specify two "index fields" in the header of each group. These fields must be numeric and monotonically increasing. Relays don't know that these fields represent, however they can service subscription requests that reference them. The streaming format would then bind index A to UTC time and index B to media time. Other formats may bind different attributes, but all could still work with generic relays.

After writing these out , I'm in favor of [3] with generic indexes.

**Comment by kixelated** [comment]
> > IMO the relay needs a way of converting relative groups to absolute groups for this to work.
> 
> If the relay can do that translation, then why couldn't the player do the same and just request an absolute group to begin with?

Yeah exactly.

UTC timestamps get annoying because of clock synchronization. It's also annoying because the player has to know the latency to the origin; requesting media that was generated 200ms ago results in a very different experience if the origin is 10 miles away versus 10k miles away.

I like the idea of delegating the group <=> timestamp mapping to the application. The catalog could contain a formula (like DASH) or it could maintain a /timeline track. That being said, we still run into the same issue; how are you supposed to request the latest group in the /timeline track? Do you always request `group=0`? And is `object=32` the latest, or do you need to wait until you receive the full /timeline?

I think there's a missing piece. If I were to implement this in HTTP, I would make a GET request with an ETAG to get the latest group/object sequence. There's no equivalent in MoQ which is problematic. I think we still need relative group requests, but the relays need to use an authoritative source somehow.

**Comment by suhasHere** [comment]
> The assumption is that the relay always has the latest group/object in cache. The reality is that it often won't, nor does it even know what the "latest" group is supposed to be.

I think that's not what it meant. By "latest", relays starts delivering whatever it has in the cache as the most recent group/object  and it doesn't need to remember anything about it other than they happen to be in cache. "waitup" is saying given everything once you get objects from current group +1.

**Comment by suhasHere** [comment]
> I think there are scalability and flexibility benefits to keeping relays agnostic to time and in fact to any internal characteristics of the objects they are serving. Keeping them decoupled allows the content to change (and the complete timing mechanism if so desired), without having to change the relay networks.
> 
> The introduction of relay’s requiring timestamp knowledge stems from the original assertion that `Neither the player nor the relay knows the mapping from group <-> timestamp.` I think this need not be true. We can teach players the relationship between group and timestamp without too much difficulty. We do this all the time with HLS and DASH. In both cases the HTTP servers delivering these formats know nothing about the relationship between segments and time, yet we are able to build players which seek to precise points in a stream.
> 
> We can do something similar with our streaming formats and the ideas already presented in this thread by Suhas and Luke. One difference is that I would make would be to allow the group offset described by Suhas to be either an absolute or relative one. As a proposal, what if the WARP catalog listed two things:
> 
> 1. The GROUP duration in milliseconds.
> 2. The name of a track which described GROUP number/timestamp pairs (exactly the track suggested by Luke above). We assume that this timeline track allows delta updates, perhaps describing the complete DVR window every minute at a group boundary, with incremental updates in between.
> 
> Use cases:
> 
> 1. **Start as close to the live edge as possible**
>    a.	Since the last group represents the last independently accessible point in the stream, the player would issue a plain SUBSCRIBE
>    b.	The relay would serve content starting with the last group it has cached (its default behavior).
>    c. If the relay does not have this content cached, and there is no pending subscription, then it would issue its own SUBSCRIBE upstream.
> 2. **Start with a 5s buffer**
>    a.	The player would calculate how many group’s it would need to guarantee a 5s buffer. Assuming the advertised group duration was 2000, then it would need 3 groups. Using zeroth based relative group addressing, it would issue SUBSCRIBE (-2).
>    b.	The relay would serve from the 3rd last group that it had cached.
>    c. If the relay does not have the content cached, it would send its own SUSBCRIBE(-2) upstream. If it has an open subscription, but has only received 2 groups, then it should wait for the 3rd before responding.
> 3. **Seek to a wall clock time of T seconds**
>    a.	The player would subscribe to the /timelines track, using SUBSCRIBE so that it receives the last full DVR description.
>    b.	Based on the data in the timeline track, the player calculates that it needs a group number of 5321. It issues a SUBSCRIBE(5321).
>    c.	The relay recognizes that this is an absolute reference to a GROUP (as all relative ones are negative) and returns content starting at GROUP #5321.
> 4. **Player encounters congestion at group 53 on 720p and wants to switch to 360p track**
>    a. WARP places the packaging restriction that group numbers across tracks must carry the same media time offset (it actually has this requirement already)
>    b. The player knows it was playing group 53 when it encountered slowness. Its next chance for a switch will be at a group boundary. Since it knows the group duration from the catalog, it can decide whether it is better off waiting for the next group, or reloading from the same group. In this case it is near the end of the group, so it decides to fetch the next group.
>    c. It issues a SUBSCRIBE (54) for the 360p track and a UNSUBSCRIBE(53) for the 720p.
> 
> Cheers Will

+1 on keeping the relays not aware of media characteristics. Since its complex and may not apply the same way to all the types of media getting transported (live vs dvr vs chat vs catalog vs something new).

+1 to keep players aware of mapping from media time to group number. I would be interested in knowing where this will not work though.  I agree with Will that the warp catalog can specify the needed relationship here

Also the susbcribe message today has "Group/Object Id" in its track request parameters. I would proposes merging Will's usecases (2) and (3) into 
    - given a Time T, find the group number you want to start fetching the media 
    - Issue a subscribe with that group number
    - Relay can decide if the requsted group number is behind or ahead of the current group and if it is in cache serve them accordingly
    Basically i think we don't need offset. But would love to hear from @wilaw  where we can't do away with absolute groups

**Comment by suhasHere** [comment]
> how are you supposed to request the latest group in the /timeline track? Do you always request `group=0`? And is `object=32` the latest, or do you need to wait until you receive the full /timeline?

We need to be clear on the term latest. In my proposal of Subscription Hints "Latest" is in terms of latest in the cache. So if the cache is current running group-45, object-32, then that's the latest.  And waitup is , wait until you see group number change before giving me objects.

I think with the hints listed, group_id & object_id being in the track request parameters and player providing the mapping function from timeline to group, we can address most of the use-cases being raised here.  I would interested to learn what use-cases cannot be addressed with the proposed scheme

**Comment by kixelated** [comment]
> > how are you supposed to request the latest group in the /timeline track? Do you always request `group=0`? And is `object=32` the latest, or do you need to wait until you receive the full /timeline?
> 
> We need to be clear on the term latest. In my proposal of Subscription Hints "Latest" is in terms of latest in the cache. So if the cache is current running group-45, object-32, then that's the latest. And waitup is , wait until you see group number change before giving me objects.

Can you give an example using the "latest" mode?.

As I understand, groups are independent but objects are not. `group=45 object=32` would be a P-frame or catalog delta. I don't understand why you would ever want to start your subscription at an undecodable object. Starting at `group=45 object=0` or `group=46 object=0` would work, but never `group=45 object=32`. Unless your objects are independently decodable, in which case they should be separate groups.

But still, the fundamental issue is basing decision making on the cache. Relays are part of a distributed system and the contents of each host's cache is highly dynamic. The behavior of "latest" would change arbitrarily based on the host without some sort of cache synchronization mechanism: something you want to avoid when scaling out.

**Comment by suhasHere** [comment]
> Starting at `group=45 object=0` or `group=46 object=0` would work, but never `group=45 object=32`.

For the video delivery case, I would use "waitup" if i have no idea what the current group is. Since it will start delivery the objects from the beginning of the next group. I would use latest for audio use-case where the `group=45 object=32` is fine for me to either render or conceal. 

Again we can enumerate the use-cases we want to address and I am happy to share my thinking around how the hints help. As part of exercise , we may come out deciding if we need more granular things than what is being proposed here.

@kixelated could you please share one liner for the use-cases you have been considering from a subscribe perspective ?

**Comment by wilaw** [comment]
> I think that's not what it meant. By "latest", relays starts delivering whatever it has in the cache as the most recent group/object and it doesn't need to remember anything about it other than they happen to be in cache.

The key point we're forgetting in the freshness quandary (or perhaps it is implicit) is that the _relay knows whether it has an open subscription for a given track_. This is a key bit of information. If it does have an open subscription, then the objects is has in cache are by definition the latest, and hence any 'relative' or 'latest' queries can be satisfied. If it does not have an open subscription, then it cannot make any determination about the freshness of any objects, even if it has some in cache and it must renew its subscription before responding. 

To illustrate this, consider a relay which has groups 637,638,639 in cache.  Five minutes later another client joins issues a SUBSCRIBE (latest). The response of the relay should be different in the two cases below:

* It has an open subscription for that track - in this case it knows that those are still the latest objects available for that track and so it would return 639
* It does not have an open subscription for that track (because the prior client that subscribed went away) - in this case the relay must make a SUBSCRIBE(latest) request upstream and return the first group it receives (perhaps 763) to the client.

> But would love to hear from @wilaw where we can't do away with absolute groups

The advantage of relative requests is that the player does not need any understanding of the relationship between group number and time. This makes for simple players, as long as they don't need to seek behind live with any accuracy. The relative datum, in this case the "live edge", is constantly changing, and for players that need to seek back accurately this can be a problem. Depending on when you make your request and how long it is queued by intermediaries, you may get two different answers. Absolute groups make precise group selection for more deterministic and are robust. For live-only players (such as the web meeting use case), we probably could get by with relative addressing alone. However since our charter includes intermediate and higher levels of latency, we should design moq-transport such that it is flexible enough to handle both relative and absolute group addressing.

**Comment by suhasHere** [comment]
> * It does not have an open subscription for that track (because the prior client that subscribed went away) - in this case the relay must make a SUBSCRIBE(latest) request upstream and return the first group it receives (perhaps 763) to the client.

if there are no upstream subscriptions active and hence no published data is being received , then there are couple of things to note
1. What's in the cache is the current source of truth and it needs to be served 
2. yes Relay may forward the subscription upstream and feed the data that arrives towards the subscriber.

This will keep "latest" property preserved. Players need to handle gaps that might be caused for several reasons including relay cache purges, publisher going away, relay reboots. I see these as player level details and not specific to moqt. Isn't it ?

**Comment by wilaw** [comment]
> What's in the cache is the current source of truth and it needs to be served

It's not a source of truth at all. It may be 5 minutes old or 50 minutes old. Without the active subscription, there is no guarantee about its status at all other than it was previously served to some client and hasn't yet fallen out of cache. Why would a client wanting "latest" want that and if it was sent, how would the client know that it wasn't the latest. It would start rendering and then jump ahead when the true latest content arrived . If you need a subscription-hint for give-me-anything-in-your-cache-not-matter-how-old-it-is, we can create that, but thats different from "latest" and I'm not clear on its utility.

**Comment by englishm** [comment]
> > I think that's not what it meant. By "latest", relays starts delivering whatever it has in the cache as the most recent group/object and it doesn't need to remember anything about it other than they happen to be in cache.
> 
> The key point we're forgetting in the freshness quandary (or perhaps it is implicit) is that the _relay knows whether it has an open subscription for a given track_. This is a key bit of information. If it does have an open subscription, then the objects it has in cache are by definition the latest, and hence any 'relative' or 'latest' queries can be satisfied. If it does not have an open subscription, then it cannot make any determination about the freshness of any objects, even if it has some in cache and it must renew its subscription before responding.
> 
> To illustrate this, consider a relay which has groups 637,638,639 in cache. Five minutes later another client joins issues a SUBSCRIBE (latest). The response of the relay should be different in the two cases below:
> 
> * It has an open subscription for that track - in this case it knows that those are still the latest objects available for that track and so it would return 639
> * It does not have an open subscription for that track (because the prior client that subscribed went away) - in this case the relay must make a SUBSCRIBE(latest) request upstream and return the first group it receives (perhaps 763) to the client.
> 

This is good insight and I think the proposed behavior makes good sense.

> > But would love to hear from @wilaw where we can't do away with absolute groups
> 
> The advantage of relative requests is that the player does not need any understanding of the relationship between group number and time. This makes for simple players, as long as they don't need to seek behind live with any accuracy. The relative datum, in this case the "live edge", is constantly changing, and for players that need to seek back accurately this can be a problem. Depending on when you make your request and how long it is queued by intermediaries, you may get two different answers. Absolute groups make precise group selection for more deterministic and are robust. For live-only players (such as the web meeting use case), we probably could get by with relative addressing alone. However since our charter includes intermediate and higher levels of latency, we should design moq-transport such that it is flexible enough to handle both relative and absolute group addressing.

In particular, I would like to see MoQ also support some improvements for DVR or even VoD playback, for example: faster frame-accurate seeking. We don't necessarily need the protocol to provide a complete implementation, but having the ability to absolutely address groups would be especially useful in that context. One concern I have about adding timestamps into the mix is in how they would apply to content being played back at arbitrary points in time relative to the original recording time.

**Comment by kixelated** [comment]
Here's my current thought.

* SUBSCRIBE can contain a relative or absolute start group.
   * A bit flag is used to specify if it's an absolute or relative request.   
   * N means serve group N and onwards
   * +1 means serve the next latest group (waitup) 
   * -0 means serve the latest group (catchup)
   * -N means serve N groups prior to latest 
* SUBSCRIBE_OK contains the latest `group=N` and `object=K`.
    * Useful for players: it knows it should start rendering after decoding object=K
    * Useful for relays: enough information to serve relative requests from cache.
* Relays can compute "latest" if it has an active subscription:
    * if SUBSCRIBE active: `latest = max(cache.latest, subscribe_ok.latest)`
    * if SUBSCRIBE inactive: `latest = new_subscribe_ok.latest`

You can serve relative subscriptions from cache only once you have an active SUBSCRIBE and the corresponding SUBSCRIBE_OK. If you don't have an active subscription, then you need to SUBSCRIBE again. The SUBSCRIBE_OK will tell the relay how stale the cache is and what can still be served.

**Comment by wilaw** [comment]
`if SUBSCRIBE active: latest = max(cache.latest, subscribe_ok.latest)`

Since cache.latest can only be equal to but never greater than subscribe_ok.latest, this simplifies to 

`if SUBSCRIBE active: latest = subscribe_ok.latest`

Another another topic, we should also allow the use case of subscribing to a stream before any groups have been produced. In this case we need some convention to convey that no groups are yet available, for example group=null and object=null

**Comment by huitema** [comment]
Please do not use these tricks of encoding meaning as the sign of an integer. It looks like a premature optimization, and in fact it is not: there is no syntax defined to encode negative integer, you will end up with a complement to 2 64 bit integer encoded on 8 bytes. It will be much cleaner and also more efficient to encode the hint as a tuple:

- meaning: an enum value describing what the subscriber wants, such as start at the current group, the next group, the first group, a specific group, or a specific location.
- additional parameters if needed, such as actual group ID if not "current", "next" or "first", and actual object ID if needs to be specified.

**Comment by ianswett** [comment]
Consensus at the interim is that we definitely want this and we both want to relative (n groups back) and range requests (ie: group n to group m), which could be open-ended.

Joining from the currently ongoing group (start_point=-1 in slides), and the next group (start_point=1) had strong interest.

start_point=0 starts from the current object within a group and there is contention about whether it's necessary given groups are intended as start points, but there was enough interest in the room that it was felt the correct way forward was to support it and then we can revisit the issue in the future if needed.

Additionally, there is interest in object granularity, particularly for the range requests.

A solution should fix #111 and #260.

**Comment by suhasHere** [comment]
+1 to christian and happy to help with the PR . thanks

**Comment by kixelated** [comment]
Here's what I was thinking:

```
BOUNDARY {
  // absolute: indexed relative to 0, increasing
  // relative: indexed relative to latest, decreasing
  // future: indexed relative to latest+1, increasing
  // We can save a byte by combining these two; ex. lower 2 bits are mode.
  Mode = none/absolute/past/future
  Sequence = varint
}

SUBSCRIBE {
  // Required: but it could be optional with default: relative/0
  Start Group:  BOUNDARY
  // Required: but it could be optional with default: absolute/0
  Start Object: BOUNDARY

  // Optional: could be a parameter, message type flag, or mode=none
  End Group:  BOUNDARY
  // Optional: could be a parameter, message type flag, or mode=none
  End Object: BOUNDARY
}
```

My preference would be `mode=none` instead of optional parameters. It's just easier to parse and encoding 0 when you don't want an end is pretty trivial and uses fewer bytes than an optional parameters table.

Here's how the modes from #289 would map:
|suhas|luke|
|-|-|
| type=StartPoint mode=PREVIOUS group=N | start_group=relative/N start_object=absolute/0 | 
| type=StartPoint mode=NEXT group=N | start_group=future/N start_object=absolute/0 |
| type=StartPoint mode=CURRENT | start_group=relative/1 start_object=absolute/0 |
| type=StartPoint mode=NOW | start_group=relative/1 start_object=relative/0 |
| type=Interval start_group=A start_object=B end_group=C end_object=D | start_group=absolute/A start_object=absolute/B end_object=absolute/C end_object=absolute/D |

I think this is simpler and more powerful. A publisher just needs to write a `resolve` function to convert a relative boundary into an absolute one based on the latest value in the cache*. This might be an asynchronous process for a relay if the cache is empty or new.

Here's some things you can't do with the proposed encoding in #289:

- Start 4 groups in the past and end 2 in the future: `start=relative/4 end=future/1`
- Start at group 69 and end at latest: `start=absolute/69 end=relative/0`
- Start at 69 and end 4 groups in the past (might be a noop): `start=absolute/69 end=relative/4`
- Keep refreshing a relative subscription while it's needed: `start=relative/4 end=future/3`
